core/dbt/parser/manifest.py

```diff
@@ -702,16 +711,28 @@ def build_manifest_state_check(self):
         # arg vars, but since any changes to that file will cause state_check
         # to not pass, it doesn't matter.  If we move to more granular checking
         # of env_vars, that would need to change.
+        # We are using the parsed cli_vars instead of config.args.vars, in order
+        # to sort them and avoid reparsing because of ordering issues.
+        stringified_cli_vars = pprint.pformat(config.cli_vars)
         vars_hash = FileHash.from_contents(
             "\x00".join(
                 [
-                    getattr(config.args, "vars", "{}") or "{}",
+                    stringified_cli_vars,
```

Comment:
Trying to give a close look at the old code versus the new code through the lens of backwards compatibility.

Can `config.cli_vars` ever be `None` or any empty string (`""`)? Or is it always a dictionary, even if an empty dictionary (`{}`)?

Just trying to convince myself that `pprint.pformat(config.cli_vars)` will be functionally equivalent to `getattr(config.args, "vars", "{}") or "{}"` in the degenerate case where `config.cli_vars` has "nothing".

Reply:
cli_vars is defined as a Dict[str, Any]. It's not an optional Dict, so it should never be None. In addition, parse_cli_vars will throw an error if it's not parsable to a dictionary.